event/addrs.go

```diff
@@ -0,0 +1,25 @@
+package event
+
+import (
+	"github.com/libp2p/go-libp2p-core/record"
+	ma "github.com/multiformats/go-multiaddr"
+)
+
+// EvtLocalAddressesUpdated should be emitted when the set of listen addresses for
+// the local host changes. This may happen for a number of reasons. For example,
+// we may have opened a new relay connection, established a new NAT mapping via
+// UPnP, or been informed of our observed address by another peer.
+type EvtLocalAddressesUpdated struct {
```

Comment:
I'm glad you called this out, because I really wasn't sure how to best approach this.

I think you're right that full snapshots are better than these diffs; in fact, I had to awkwardly keep a reference to the Host in the consumer code in go-libp2p, since it needs the full current state to produce the PeerRecord, so the diff was being produced and then ignored...

I like option four, maybe with the removed addresses in a separate list, which would let you ignore them completely if you just want current addrs:

```
type AddrAction int
const (
  Unknown AddrAction = iota
  Added
  Maintained
  Removed
)

type UpdatedAddress struct {
  ma.Multiaddr
  Action AddrAction
}

type EvtLocalAddressesUpdated struct {
  CurrentAddrs []UpdatedAddress
  Removed []UpdatedAddress
}
```

The above uses an `Unknown` hint for producers that only have the current state and can't produce the diff hints. With that, do you think we still need a flag for whether hints are available?

Reply:
That implementation proposal is good. My gut feeling tells me to replace the slices with `map[ma.Multiaddr]AddrAction`, so that the consumer can index by known multiaddrs, but that's worthless because `ma.Multiaddr` is actually an interface and therefore what's stored in the key is a pointer, which will be different every time. Indexing by a pointer in this case is useless.

We might still need to know if the publisher is sending hints. Otherwise, we can't disambiguate between the publisher not tracking at all, or `Removed` just being empty in this case because no addresses were removed. WDYT?

Reply:
That makes sense - the flag would be useful for the `Removed` case.

We could do a map of hints if we used the `string(addr.Bytes())` as the keys and added a helper method to the event type. So the caller could do e.g. `evt.GetAction(addr)`. Then we wouldn't need the `UpdatedAddress` type, which would be nice for consumers that don't care about the hints at all.

Reply:
That would work, but in all honesty, it's sounding a bit convoluted at this point. I'd just add a third `Diffs bool` field:

```go
type AddrAction int

const (
  Unknown AddrAction = iota
  Added
  Maintained
  Removed
)

type UpdatedAddress struct {
  Address ma.Multiaddr
  Action AddrAction
}

type EvtLocalAddressesUpdated struct {
  Diffs        bool
  Current      []UpdatedAddress
  Removed      []UpdatedAddress
}
```